File: src/elixir/svd/lib/svd/exporters/formats/json_exporter.ex
defmodule SVD.Exporters.Formats.JSONExporter do
  @moduledoc false

  alias SVD.Exporters.{BaseExporter, Helpers, StandardisedVesselDatasetContent}
  alias SVD.Serializers
  alias SVD.Validators.StandardisedVesselDatasetValidator

  defstruct validator: StandardisedVesselDatasetValidator

  @type t :: %__MODULE__{validator: module() | nil}

  @spec new(module() | nil) :: t()
  def new(validator_module), do: %__MODULE__{validator: validator_module}

  @spec export_async(t(), term(), term()) ::
          {:ok, StandardisedVesselDatasetContent.t()} | {:error, term()}
  def export_async(%__MODULE__{validator: validator_module}, svd, _ctx \\ nil) do
    BaseExporter.validate_and_export(validator_module, svd, fn dataset ->
      case Jason.encode(Serializers.to_json_map(dataset), pretty: true) do
        {:ok, json} ->
          {:ok,
           StandardisedVesselDatasetContent.new(
             json,
             Helpers.file_name(dataset, "json"),
             "application/json"
           )}

        {:error, error} ->
          {:error, error}
      end
    end)
  end

  @spec decode(binary()) :: {:ok, SVD.Models.StandardisedVesselDataset.t()} | {:error, term()}
  def decode(payload) when is_binary(payload) do
    with {:ok, json_payload} <- Jason.decode(payload),
         {:ok, dataset} <- Serializers.from_json_map(json_payload) do
      {:ok, dataset}
    end
  end

  def decode(_), do: {:error, :invalid_json_payload}
end

File: src/go/svd/exporters/formats/json/json_exporter.go
package json

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/models"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/validators"
)

var marshalIndentFunc = json.MarshalIndent

// Exporter implements the JSON exporter for SVD.
type Exporter struct {
	*exporters.BaseExporter
}

// NewJSONExporter creates a new JSON exporter with the given validator.
func NewJSONExporter(validator validators.SVDValidator) exporters.SVDJSONExporter {
	return &Exporter{BaseExporter: exporters.NewBaseExporter(validator)}
}

// ExportAsync exports the SVD to JSON format.
func (e *Exporter) ExportAsync(ctx context.Context, svd *models.StandardisedVesselDataset) (*exporters.StandardisedVesselDatasetContent, error) {
	return e.ValidateAndExport(ctx, svd, e.internalExport)
}

func (e *Exporter) internalExport(ctx context.Context, svd *models.StandardisedVesselDataset) (*exporters.StandardisedVesselDatasetContent, error) {
	data, err := marshalIndentFunc(svd, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("marshal svd json: %w", err)
	}

	return &exporters.StandardisedVesselDatasetContent{
		Data:        data,
		FileName:    generateFileName("json", svd),
		ContentType: "application/json",
	}, nil
}

func generateFileName(ext string, svd *models.StandardisedVesselDataset) string {
	if svd != nil && svd.General != nil {
		return fmt.Sprintf("SVD_%s_%s.%s", svd.General.Imo, svd.General.ShipReportingDate.UTC().Format("2006-01-02"), ext)
	}
	return fmt.Sprintf("SVD_%s.%s", time.Now().UTC().Format("2006-01-02"), ext)
}
